AX: Make AXIsolatedObject::m_childrenIDs a FixedVector to save 8-bytes per object, and remove children IDs that were hydrated into m_children to save even more memory

twilco opened the pull request:
#### 3173e6ac4f4b9cf348687fb7a8c002aa1d75f05d
<pre>
AX: Make AXIsolatedObject::m_childrenIDs a FixedVector to save 8-bytes per object, and remove children IDs that were hydrated into m_children to save even more memory
<a href="https://bugs.webkit.org/show_bug.cgi?id=296509">https://bugs.webkit.org/show_bug.cgi?id=296509</a>
<a href="https://rdar.apple.com/156749138">rdar://156749138</a>

Reviewed by Joshua Hoffman.

FixedVector is 8-bytes smaller than Vector, so using that for m_childrenIDs (now called m_unresolvedChildrenIDs) saves
8 bytes per object. This necessitated a rewrite of handling m_pendingSubtreeRemovals in AXIsolatedTree::applyPendingChangesLocked,
as the old implementation leaned on the fact that m_childrenIDs was a Vector.

This commit also further saves memory by eliminating AXIDs from m_childrenIDs (now called m_unresolvedChildrenIDs) as
they are resolved into actual objects and placed in AXIsolatedObject::m_children. The effect of this is smaller than I
had hoped, only saving 0.5mb on gmail.com, but I believe it&apos;s still an improvement worth taking.

* Source/WebCore/accessibility/isolatedtree/AXIsolatedObject.cpp:
(WebCore::AXIsolatedObject::AXIsolatedObject):
(WebCore::AXIsolatedObject::detachRemoteParts):
(WebCore::AXIsolatedObject::setChildrenIDs):
(WebCore::AXIsolatedObject::children):
* Source/WebCore/accessibility/isolatedtree/AXIsolatedObject.h:
* Source/WebCore/accessibility/isolatedtree/AXIsolatedTree.cpp:
(WebCore::AXIsolatedTree::queueRemovalsLocked):
(WebCore::AXIsolatedTree::applyPendingChangesLocked):
* Source/WebCore/accessibility/isolatedtree/AXIsolatedTree.h:

Canonical link: <a href="https://commits.webkit.org/297897@main">https://commits.webkit.org/297897@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/ae0808732ce12c06bb18559f51a1ffa276320b82

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/113260 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/33000 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/23473 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/119468 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/64007 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/115222 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/33648 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/41558 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/86199 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/41268 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/116207 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/26876 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/101880 "Passed tests") | [❌ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/66523 "Found 130 new API test failures: /WPE/TestWebsiteData:/webkit/WebKitWebsiteData/configuration, /WPE/TestBackForwardList:/webkit/WebKitWebView/navigation-after-session-restore, /WPE/TestInputMethodContext:/webkit/WebKitInputMethodContext/reset, /WPE/TestInputMethodContext:/webkit/WebKitInputMethodContext/surrounding, /WPE/TestLoaderClient:/webkit/WebKitURIResponse/http-headers, /WPE/TestWebKitWebContext:/webkit/WebKitWebContext/timezone, /WPE/TestWebKitWebContext:/webkit/WebKitWebContext/timezone-worker, /WPE/TestWebKitUserContentManager:/webkit/WebKitUserContentManager/injected-script, /WPE/TestLoaderClient:/webkit/WebKitWebView/unfinished-subresource-load, /WPE/TestWebsiteData:/webkit/WebKitWebsiteData/cookies ... (failure)") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/26141 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/20008 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/63222 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/96251 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/20083 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/122685 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/40338 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/30093 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/95052 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/40723 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/98095 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/94798 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/24196 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/39940 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/17755 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/36474 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/40224 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/45723 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/39865 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/43198 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/41602 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/bcee2b2055c882db71c4c45fb635165a699757cf. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/48548

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/ae0808732ce12c06bb18559f51a1ffa276320b82. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/48548

twilco committed:
AX: Make AXIsolatedObject::m_childrenIDs a FixedVector to save 8-bytes per object, and remove children IDs that were hydrated into m_children to save even more memory
https://bugs.webkit.org/show_bug.cgi?id=296509
rdar://156749138

Reviewed by Joshua Hoffman.

FixedVector is 8-bytes smaller than Vector, so using that for m_childrenIDs (now called m_unresolvedChildrenIDs) saves
8 bytes per object. This necessitated a rewrite of handling m_pendingSubtreeRemovals in AXIsolatedTree::applyPendingChangesLocked,
as the old implementation leaned on the fact that m_childrenIDs was a Vector.

This commit also further saves memory by eliminating AXIDs from m_childrenIDs (now called m_unresolvedChildrenIDs) as
they are resolved into actual objects and placed in AXIsolatedObject::m_children. The effect of this is smaller than I
had hoped, only saving 0.5mb on gmail.com, but I believe it's still an improvement worth taking.

* Source/WebCore/accessibility/isolatedtree/AXIsolatedObject.cpp:
(WebCore::AXIsolatedObject::AXIsolatedObject):
(WebCore::AXIsolatedObject::detachRemoteParts):
(WebCore::AXIsolatedObject::setChildrenIDs):
(WebCore::AXIsolatedObject::children):
* Source/WebCore/accessibility/isolatedtree/AXIsolatedObject.h:
* Source/WebCore/accessibility/isolatedtree/AXIsolatedTree.cpp:
(WebCore::AXIsolatedTree::queueRemovalsLocked):
(WebCore::AXIsolatedTree::applyPendingChangesLocked):
* Source/WebCore/accessibility/isolatedtree/AXIsolatedTree.h:

Canonical link: https://commits.webkit.org/297897@main

webkit-commit-queue commented:
Committed 297897@main (3173e6ac4f4b): <https://commits.webkit.org/297897@main>

Reviewed commits have been landed. Closing PR #48548 and removing active labels.
